internal/controllers/gateway/route_utils.go

```diff
@@ -897,3 +897,51 @@ func isParentRefEqualToParent[
 
 	return true
 }
+
+func isRouteAcceptedByListener[T types.RouteT](ctx context.Context,
+	mgrc client.Client,
+	route T,
+	listener gatewayv1beta1.Listener,
+	gateway gatewayv1beta1.Gateway,
+	parentRef gatewayv1beta1.ParentReference,
+) (bool, error) {
+	// Check if the route matches listener's AllowedRoutes.
+	if ok, err := routeMatchesListenerAllowedRoutes(ctx, mgrc, route, listener, gateway.Namespace, parentRef.Namespace); err != nil {
+		return false, fmt.Errorf("failed matching listener %s to a route %s for gateway %s: %w",
+			listener.Name, route.GetName(), gateway.Name, err,
+		)
+	} else if !ok {
+		return false, nil
+	}
+
+	// Check the listeners statuses:
+	// - Check if a listener status exists with a matching type (via SupportedKinds).
+	// - Check if it matches the requested listener by name (if specified).
+	// - And finally check if that listeners is marked as Ready.
+	if err := existsMatchingReadyListenerInStatus(route, listener, gateway.Status.Listeners); err != nil {
+		// return no error here, as we don't care of the reason why this check failed.
+		return false, nil //nolint:nilerr
+	}
+
+	// Check if listener name matches.
+	if parentRef.SectionName != nil && *parentRef.SectionName != "" && *parentRef.SectionName != listener.Name {
+		return false, nil
+	}
+
+	// Perform the port matching as described in GEP-957.
+	if parentRef.Port != nil && *parentRef.Port != listener.Port {
+		// This ParentRef has a port specified and it's different
+		// than current listener's port.
+		return false, nil
+	}
+
+	if !routeTypeMatchesListenerType(route, listener) {
+		return false, nil
+	}
+
+	if !routeHostnamesIntersectsWithListenerHostname(route, listener) {
+		return false, nil
+	}
+
+	return true, nil
+}
```

Comment:
Could use some unit tests

Reply:
This function is pretty simple: just a series of calls to other functions and a check on the return values. I think that adding unit tests for the functions called therein is way more beneficial but out of scope for this PR. I can create an issue to improve testing in that sense, WDYT?

Reply:
I get that the function is relatively simple. I am however of the general opinion that we should add unit test coverage as much as is feasible. There can be several benefits to this, including but not limited to:

* instigating a deeper review of the logic that we've written, helping to highlight areas for improvement
* creating codified contracts of how we expect the functionality to work, simple or not, for future readers and maintainers
* helping future developers avoid accidental regressions or unexpected breakages

While it's probably technically true that for this _current function_ the benefits are not as great as say, some other functions in the vicinity I still prefer a "test focused mindset": I feel the praxis of prioritizing tests has general benefits to both the coder and the codebase and is enriching.

This is not a blocking comment, but I do hope that I've managed to be at least a bit persuasive and that I'm not coming across as arbitrary.

Reply:
I get your point, and it makes sense. I'll add a test for that 👍

Reply:
here it is 🙂  374db328b5d7bf6ea39687649e44ace168f2a104